LLEXT: support detached sections

lyakh opened the pull request:
SOF loads LLEXT objects into slower IMR memory but they're partially pre-linked for SRAM, where objects will be copied to after linking by LLEXT core. But some sections should not be copied to SRAM and be used in IMR directly. This PR adds support for such sections.

lyakh committed:
llext: make llext_loaded_sect_ptr() available to all LLEXT code

Make llext_loaded_sect_ptr() a proper function to be able to call it
from platform LLEXT code.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: (cosmetic) use a more common name for a variable

It's more common in the LLEXT code base to call elf_sym_t instances
"sym" and not "sym_tbl." Convert llext_link_plt() to follow this
convention.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
LLEXT: Xtensa: add support for L32R relocation

When building LLEXT for Xtensa with custom sections the compiler
can leave unresolved references in them. Then this has to be done by
the LLEXT core during linking. This commit adds linking support for
the L32R Xtensa instruction.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

pillo79 commented on subsys/llext/llext_load.c:
```diff
@@ -274,6 +287,16 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext)
+		 * sections that will be placed in slower memory. Don't merge such
+		 * sections into main regions
+		 */
+		if (pre_located && shdr->sh_addr < LLEXT_SECTION_MIN_ADDR) {
+			continue;
+		}
```
Why do addresses below a certain value get ignored?
I think this is very project-specific. Should we have callback functions to provide this kind of customizations?

lyakh replied:
this is basically a `NULL` area. I use 0 as a base for sections without pre-defined run-time location, but since there can be several such sections, some of them get other low addresses

pillo79 commented on subsys/llext/llext_load.c:
```diff
@@ -574,7 +597,10 @@ static int llext_copy_symbols(struct llext_loader *ldr, struct llext *ext,
+			const char *s_name = llext_string(ldr, ext, LLEXT_MEM_SHSTRTAB,
+							  shdr->sh_name);
+
+			if (pre_located && section_addr >= LLEXT_SECTION_MIN_ADDR) {
```
This also feels specific - although maybe a per-section flag could work (I assume they are the sections _not_ ignored by the above check).

pillo79 reviewed (commented):
IMO some parts of this are very project specific. To improve this situation, I would suggest to pass the `struct llext_load_params` around, instead of expanding its fields, and add more fields and/or callbacks to that.

pillo79 commented on subsys/llext/llext_load.c:
```diff
@@ -36,8 +36,21 @@ LOG_MODULE_DECLARE(llext, CONFIG_LLEXT_LOG_LEVEL);
  *   calls, taking into account if the load process was successful or not.
  */
 
+#define LLEXT_SECTION_MIN_ADDR 0x10000
```
This hardcoded value which is neither arch- nor even project- independent. If comparing addresses is required, then this must at least be customizable by the caller.

lyakh replied:
@pillo79 agree, yes, I was thinking about it too. Would it be ok for you to pass this as a value in `struct llext_load_param`? Otherwise we could add a platform callback to that structure to perform any testing and return `true` or `false` for whether the symbol is used where it's located or where it's linked for, but I think that would make it uglier, less transparent.

pillo79 replied:
I personally see it as the opposite - it would neatly decouple the generic implementation from project-specific reasoning.
For example, I can already see places where I would add callbacks for: 
* "do we allocate a copy of a section?" <- would replace and extend `STORAGE_WRITABLE`, e.g. allow `rodata` from Flash
* "how do we relocate a section?" <- would extend (or even replace!) `pre_located` and simplify the current PR
* "what is the effective base address of this section?" <- would possibly replace link scripts and also be a first step for in-Flash XIP.

lyakh committed:
llext: pass the whole struct llext_load_param

To simplify LLEXT loader parameter extension pass the whole
struct llext_load_param to final consumers instead of extracting
individual fields from it in the caller function.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
LLEXT: add support for detached sections

Some LLEXT objects can include sections, that should not be merged
with other sections of the same type. E.g. when such sections should
be placed into locations, other than the default. Add support for
such sections.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

pillo79 reviewed (approved):
This looks awesome, and the basis for other improvements! Thanks!
